src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Split.cs

```diff
@@ -0,0 +1,78 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    public static partial class MemoryExtensions
+    {
+        public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span)
+            => new SpanSplitEnumerator<char>(span, ' ');
+
+        public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> span, char separator)
+            => new SpanSplitEnumerator<char>(span, separator);
+
+        public static SpanSplitSequenceEnumerator<char> Split(this ReadOnlySpan<char> span, string separator)
+            => new SpanSplitSequenceEnumerator<char>(span, separator);
+    }
+
+    public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
+    {
+        private readonly ReadOnlySpan<T> _sequence;
+        private readonly T _separator;
+        private int _offset;
+        private int _index;
+
+        public SpanSplitEnumerator<T> GetEnumerator() => this;
+
+        internal SpanSplitEnumerator(ReadOnlySpan<T> span, T separator)
+        {
+            _sequence = span;
+            _separator = separator;
+            _index = 0;
+            _offset = 0;
+        }
+
+        public Range Current => new Range(_offset, _offset + _index - 1);
+
+        public bool MoveNext()
+        {
+            if (_sequence.Length - _offset < _index) { return false; }
+            var slice = _sequence.Slice(_offset += _index);
+
+            var nextIdx = slice.IndexOf(_separator);
+            _index = (nextIdx != -1 ? nextIdx : slice.Length) + 1;
+            return true;
+        }
+    }
+
+    public ref struct SpanSplitSequenceEnumerator<T> where T : IEquatable<T>
+    {
+        private readonly ReadOnlySpan<T> _sequence;
+        private readonly ReadOnlySpan<T> _separator;
+        private int _offset;
+        private int _index;
+
+        public SpanSplitSequenceEnumerator<T> GetEnumerator() => this;
+
+        internal SpanSplitSequenceEnumerator(ReadOnlySpan<T> span, ReadOnlySpan<T> separator)
+        {
+            _sequence = span;
+            _separator = separator;
+            _index = 0;
+            _offset = 0;
+        }
+
+        public Range Current => new Range(_offset, _offset + _index - 1);
+
+        public bool MoveNext()
+        {
+            if (_sequence.Length - _offset < _index) { return false; }
+            var slice = _sequence.Slice(_offset += _index);
```

Comment:
Wouldn't you need to do something along these lines here to account for the length of the separator? Otherwise you're including part of the separator in your output...

`var slice = _sequence.Slice(_offset += _index + (_index == 0 ? 0 : _separator.Length - 1));`

Consider a unit test like this one:

```
[Theory]
[InlineData("aaababbaaabbabbaba", "bb")]
[InlineData("aaaa\r\nbbb\r\ncccc\ncc\r\ndddd", "\r\n")]
public void SplitAllChars(string str, string separator)
{
    var parts = new List<string>();

    var span = str.AsSpan();
    foreach (var range in span.Split(separator))
        parts.Add(span[range].ToString());

    Assert.Equal(str.Split(new[] { separator }, StringSplitOptions.None), parts);
}
```

Reply:
Thank you for bringing this to my attention; in the latest commit where I've merged the two enumerator implementations I've resolved this issues.